cpp/src/arrow/dlpack.h

```diff
@@ -0,0 +1,32 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include "arrow/array/array_base.h"
+#include "arrow/dlpack_structure.h"
+
+namespace arrow {
+
+/// \brief DLPack protocol for producing DLManagedTensor
+///
+/// Returns pointer to the DLManagedTensor class defined by
+// the DLPack protocol
+ARROW_EXPORT
+DLManagedTensor* ExportToDLPack(const std::shared_ptr<Array>& arr);
```

Comment:
It seems fragile to return a raw pointer. Instead, you could probably take an input pointer allocated by the caller. cc @kkraus14 for confirmation.
```suggestion
Status ExportArray(const std::shared_ptr<Array>& arr, DLManagedTensor* out);
```

Reply:
Looks like there's a mix of how other projects do it. Returning a raw pointer:
* https://github.com/rapidsai/cudf/blob/cfa2d513667edabda6c4487f15f251f757f0c94d/cpp/include/cudf/interop.hpp#L88
* https://github.com/pytorch/pytorch/blob/d6578b36781b0402163e230e8838a802bf51ba0f/aten/src/ATen/DLConvertor.h#L14
* https://github.com/tensorflow/tensorflow/blob/44e4863a120282b43f4f2e4eca4312935c881522/tensorflow/c/eager/dlpack.h#L28

Taking an input pointer allocated by the caller:
* https://github.com/apache/tvm/blob/bce82432f98a93356f393166ae50834cb177caa9/include/tvm/runtime/c_runtime_api.h#L562

I think @pitrou's proposal here makes sense.

Reply:
The semantics of a raw pointer are not obvious to say the least. Who is responsible to deallocate it and how?

Reply:
I think this should do: https://github.com/apache/arrow/pull/38472/commits/b3c0a8c8fe73d61c241c797e851384d623455c1c

Reply:
@AlenkaF I had missed this was in the spec. This seems ok on the principle. Feel free to ping me when there's new code to review :-)

Reply:
Will do, thanks!! (Just finishing up the docs and then I will mark the PR ready for review and add a ping ❤️ )